### ISSUE TITLE
Testing Series Question (Video 8: reduce, deciphering docs)

### ISSUE THREAD

**Issue by nikitanemtcev** [gordonmzhu/questions#295]
Hi Gordon,
### What is the associated Watch and Code lesson? 

The associated Watch and Code lessons are [8. reduce, deciphering docs](https://watchandcode.com/courses/77710/lectures/1484439) and [9. reduce, finishing up.](https://watchandcode.com/courses/77710/lectures/1513011)

I encountered something confusing in part 8 at the 13:00 minute mark.

I was confused because I don't understand how to approach testing.

For example, one of the issues I stumbled upon is whether to start by reading the MDN docs.

If I recall correctly you mentioned in a different lesson that you should be able to write the tests without looking at the docs even once as a way to gauge whether you understand testing.

How do I know whether to read the docs first, write requirements, then write tests, or to start with writing tests right away and not read the docs at all?

I gave this a bit of thought: is it a matter of deciding how complicated the method is that you're trying to replicate, and adjusting your method accordingly (e.g. for more complex methods such as reduce, it's important to read the docs first)?

### Include a live demo link so that I can run your code. 

I've uploaded my incomplete test code to GitHub at [a repository here.](https://github.com/nikitanemtcev/testing-series-question/blob/master/reduce.html)

In this repository, there are only requirements written in the comments of the code which would form the basis for the test code.

### Explain the exact steps I need to take to replicate the problem. 

To provide some more context for my issue, I wrote two insights that I captured while watching part 1 of the copyWithin series with Jake Ryan. I used these two insights as rough guidelines for writing my requirements.

1) Use code examples to write your tests.

2) Think about what's the next simplest example that will force you to write test code.

My approach to testing is as follows:

- I didn't read the MDN docs at all. I started by writing code in the console to test how the native reduce method works.
- I started with the simplest code examples I could think of, 'simplest' meaning writing as few lines of code as possible and with the fewest number of arguments in the functions. I wrote code using the native array method by starting with an array with one element. I copied the code to Sublime Text when the code seemed to demonstrate something significant.

The simplest test code that I could think of to start with is the following:

```
reduce([1], function(previousValue) {
   eq(previousValue, 1);
});
```

I thought that this code was significant because the value of the first parameter in the callback function was the solo element in the array. This happened only when the array had one element and no initialValue was provided, so I wrote a test for it.

However, at 13:00 the simplest test that you start with is testing whether previousValue is equal to initialValue. The test code that you have for this is:

```
reduce([1], function(previousValue) {
   eq(previousValue, 0);
}, 0);
```

It seems that even though I tried to start with the simplest code example, my first requirement and test was completely different from yours and I'm curious as to how you determine how simple/complex a test is.

In summary, my questions are:

1) How do you know whether to start with reading the MDN docs?
2) My first requirement/test is completely different from yours. What is your criteria for deciding whether a code example/test is the simplest that you can think of?
3) When you say that you should be able to write tests without looking at the docs even once, why would this be necessary? Wouldn't it be quite difficult to figure out the basic idea/functionality of a certain method, which would make it difficult to write tests?

Best regards,

Nikita

**Comment by gordonmzhu** [comment]
**Response to #1 and #3**

In this situation, the documentation is a good starting point because it can give you an idea about how things should generally work. However, when the documentation is confusing, incomplete, or absent (which is quite common), you should be able to experiment with things on your own to figure out how they work (you can think of this as reverse engineering). So when documentation is available, use it, but don't rely on it completely.

**Response to #2.**

Your "simplest case" and mine are pretty similar. The only difference is that I supplied an initial value and you didn't. Something like this is borderline and there are reasonable arguments on both sides about which is simpler. What do you think? 

In these gray areas, it's not important that you get it "right". What's important is that you generally start simple and then slowly build up. It's possible that two reasonable programmers might two things slightly differently, and that's ok.
